crates/factors/src/lib.rs

```diff
@@ -22,10 +22,55 @@ pub type Linker<T> = wasmtime::component::Linker<<T as RuntimeFactors>::Instance
 pub type App = spin_app::App<'static, spin_app::InertLoader>;
 pub type AppComponent<'a> = spin_app::AppComponent<'a, spin_app::InertLoader>;
 
-// TODO: Add a real Error type
-pub type Result<T> = wasmtime::Result<T>;
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("factor dependency ordering error: {0}")]
+    DependencyOrderingError(String),
+    #[error("no such factor: {0}")]
+    NoSuchFactor(&'static str),
+    #[error("{factor} requested already-consumed key {key:?}")]
+    RuntimeConfigReusedKey { factor: &'static str, key: String },
+    #[error("runtime config error: {0}")]
+    RuntimeConfigSource(#[source] anyhow::Error),
+    #[error("unused runtime config key(s): {}", keys.join(", "))]
+    RuntimeConfigUnusedKeys { keys: Vec<String> },
+    #[error("{factor} {method} failed: {source}")]
+    RuntimeFactorError {
```

Comment:
It might ultimately be better to split this out into 4 separate errors e.g. `FactorInitError`, `FactorPrepareError`.

Reply:
I would certainly be in favor of this!